Fix: WebSocket SubscribeResponse not being delivered to clients

sanity opened the pull request:
## Summary
- Fixes WebSocket SubscribeResponse delivery to clients in River integration
- Handles local subscriptions when no remote peers are available  
- Ensures completed operations flow through client notification system

## Problem Description
The River integration tests were hanging indefinitely while waiting for SubscribeResponse messages that were never delivered via WebSocket. This occurred in two specific scenarios:

1. **No remote peers available**: When attempting to subscribe to a contract with no remote peers available, the operation would fail immediately without checking if the contract existed locally
2. **Operation completion at originating node**: When subscribe operations completed at the originating node, they weren't properly flowing through the client notification system

## Root Cause
The subscribe operation logic had two gaps:

1. **Missing local subscription handling**: When `ring.closest_potentially_caching()` returned `None` (no remote peers), the code immediately returned an error without checking if the contract was available locally
2. **Incomplete operation completion flow**: Operations completing at the originating node weren't being properly pushed back through the operation manager to trigger client notifications

## Solution Implemented
1. **Local subscription support**: When no remote peers are available, check if the contract exists locally and create a local subscription if it does
2. **Proper operation completion**: Ensure completed operations are pushed back through the operation manager so they get reported to clients via WebSocket
3. **Enhanced logging**: Added detailed logging to help debug subscription flow and completion

## Testing
- Verified with River integration tests - subscription operations now complete successfully
- Local subscriptions work when no remote peers are available
- WebSocket SubscribeResponse messages are properly delivered to clients
- No regressions in existing subscription functionality

## Files Changed
- `crates/core/src/operations/subscribe.rs`: Core subscription logic improvements

Fixes https://github.com/freenet/river/issues/33

[AI-assisted debugging and comment]

iduartgomez commented on crates/core/src/operations/subscribe.rs:
```diff
@@ -71,9 +71,39 @@ pub(crate) async fn request_subscribe(
+                if super::has_contract(op_manager, *key).await? {
+                    // We have the contract locally, subscribe to ourselves
+                    let own_location = op_manager.ring.connection_manager.own_location();
+                    if op_manager
```
I am not sure why we need to do this even... we may as well complete the operation. As with all other ops, having self as target for network ops leads to problems more often than not.

I think that this should already be handled by the ws API and all the client request routing and the Session Actor machinery

sanity commented:
Updated based on review feedback:

## Changes Made
Removed the self-subscription logic as requested. The fix now:
- ✅ Detects when contract is available locally (when no remote peers available)  
- ✅ Marks the operation as completed (so client gets SubscribeResponse)
- ✅ Does NOT add the node as a network subscriber to itself
- ✅ Lets contract_handler manage local subscriptions separately

## Key Change
Removed these lines:
```rust
let own_location = op_manager.ring.connection_manager.own_location();
if op_manager.ring.add_subscriber(key, own_location.clone()).is_err() {
    // ...
}
```

The operation now simply completes when the contract is locally available, maintaining proper separation between network-level and application-level subscriptions.

[AI-assisted implementation]

sanity commented:
## Test Results After Update

I've re-run the River integration test with the updated fix (self-subscription removed). Unfortunately, the test appears to have regressed:

### Test Status
- ❌ Test times out after 120 seconds
- ❌ Test hangs during initial setup (before even reaching subscription step)
- ❌ No subscription-related log messages observed

### Details
The test () compiled successfully but hung during the initial node setup phase. It didn't progress far enough to test the subscription functionality. The last output shows:
- Node locations assigned (Gateway, Alice, Bob, Charlie)
- Initial state configured
- Then the test hangs without starting the actual test steps

### Possible Issues
1. The removal of self-subscription may have introduced a deadlock when nodes have no peers
2. There might be an issue with how local contracts are handled when no remote peers exist
3. The operation completion without actual subscription might be causing issues in the test setup

### Recommendation
The fix may need additional investigation to understand why the test hangs so early. The issue appears to be more fundamental than just the Subscribe operation - possibly related to how nodes initialize when they can't find peers to subscribe to.

Note: Nacho's comment about self-subscription being problematic is validated, but we may need a different approach to handle the local-only scenario properly.

[AI-assisted testing and analysis]

iduartgomez commented:
My guess is 

> There might be an issue with how local contracts are handled when no remote peers exist

may be the issue.

sanity commented:
@claude Can you check whether this has been fixed already per @iduartgomez's question above? If so please close this PR.

sanity commented:
@claude please resolve conflicts and ensure PR is up-to-date

iduartgomez commented:
After resolving the merge conflicts, the issue may be solved cause we actually pulled the changes from main... try to run the tests with the fixes in main only and see if they still do not pass.

iduartgomez commented:
@claude there are compile issues with the tests and formatting issues

iduartgomez commented:
@claude there still are compile issues with isolated_node_regression

sanity commented:
[Codex] Thanks for surfacing this. The op state machines expect every completion (even local) to flow through `handle_op_result(...)` so the event loop can call `report_result(...)`. Pushing a completed op back into the manager skips that step, which is why the WebSocket never gets the `SubscribeResponse`.

The pattern we already use in `put::request_put` when we finish locally (see `crates/core/src/operations/put.rs:965`) is: (1) move the op into the state that the state machine expects (`AwaitingResponse { upstream: None, ... }`), (2) synthesize the protocol message that represents the next transition, and (3) inject it with `op_manager.notify_op_change(...)`. When `process_message` handles that synthetic message it produces an `OperationResult` with `return_msg = None` plus a finalized state, which triggers the `finalized()` guard in `handle_op_result` (`crates/core/src/operations/mod.rs:105-113`) and in turn hits `report_result(...)`.

For subscribe we can do the same without a special local-only bypass: after we detect the contract locally, set the op state to the usual `AwaitingResponse` variant (skip list empty, retries 0, `upstream_subscriber: None`, `current_hop` as usual) and synthesize a loopback `SubscribeMsg::ReturnSub { subscribed: true }` using our own location as both `sender` and `target`. Inject that via `notify_op_change(NetMessage::from(loopback), OpEnum::Subscribe(...))`. Because `upstream_subscriber` is `None`, the `ReturnSub` branch in `subscribe.rs` (`crates/core/src/operations/subscribe.rs:410-460`) will mark the op as `Completed` and leave `return_msg = None`, so the event loop immediately goes through `report_result` and delivers `SubscribeResponse` to the waiting client.

One extra tweak while doing this: guard the `ring.add_subscriber` call in that branch so we skip it when `sender.peer == own.peer`; that keeps us from reintroducing the self-subscriber regression Nacho pointed out earlier, while still allowing the loopback completion.

iduartgomez commented:
Usually we either artificially create a state and transit to it to trigger whatever path, or, if posible just use ‘notify_node_change’ and handle that at the p2p event loop.

iduartgomez commented:
Also qe dont need to solve for legacy mode, since we are removing it soon

sanity commented:
@iduartgomez I've implemented your  suggestion and the message successfully flows through the entire pipeline:

1. ✅ Subscribe operation detects local contract and calls 
2. ✅ P2P event loop receives the event and sends SubscribeResponse to result router
3. ✅ Result router receives it and forwards to session actor
4. ❌ Session actor can't deliver - transaction not registered

**Root cause discovered:** Transaction mismatch in session actor. The session actor only delivers results to clients that have been registered via `RegisterTransaction`. 

Looking at the logs:
- Request router creates Subscribe operation with transaction `01K6E690C7YTRHAM4AKCRCYB83`
- Subscribe operation sends LocalSubscribeComplete with transaction `01K6E690C70E6VZM58TWTYD483` (different!)
- Session actor has zero registered transactions for the Subscribe transaction ID

The Subscribe operation appears to be using a different transaction ID than what was registered when the client made the request. This means the session actor can't correlate the response back to the waiting client.

**Questions:**
1. Should Subscribe operations register their transaction with the session actor before completing locally?
2. Or should we look up the registered transaction from the original client request and use that ID instead?
3. Is there a mapping between operation transaction IDs and client request transaction IDs I should be using?

The actor-based client management system needs the transaction ID to be consistent throughout the request/response cycle, but I'm not sure where this divergence is happening or how to fix it properly.

[AI-assisted debugging and comment]

iduartgomez commented:
I think we can be a bit smarter for the routing of requests/responses, since if you are doing a subscribe , you dont particularly about what transaction completes it, but if the given contract has been or not subscribed. No matter what transaction, the result is idempotent. The same happens with gets. Only puts and updates need to be tracked exactly per each transaction.

sanity commented:
@iduartgomez Thanks for the guidance! I understand the philosophical point that Subscribe/Get are idempotent so exact transaction tracking isn't critical.

However, I'm unclear on **how** to implement "Subscribe operations should register their transaction with the session actor before completing locally."

**The architectural challenge:**
- Subscribe operations execute in the operations layer (`operations/subscribe.rs`)
- Transaction registration happens in the client request handling layer (when requests come from WebSocket)
- `OpManager` has access to `result_router_tx` but NOT to the session actor's `SessionMessage` channel

**Current flow for normal operations:**
1. Client request → WebSocket handler → Request router creates operation
2. Request router sends `RegisterTransaction` to session actor
3. Operation completes → sends result to result_router → result_router forwards to session actor
4. Session actor looks up registered clients and delivers response

**The problem with local Subscribe:**
- When Subscribe detects local contract, it completes immediately
- But it completes BEFORE the request router can register the transaction
- So session actor has no record of which client to deliver to

**Question:** Should I:
1. Add a `session_actor_tx` field to `OpManager` so operations can register themselves?
2. Or change the request router to register transactions BEFORE starting operations (instead of after)?
3. Or use a different approach entirely?

The current `notify_node_event` implementation successfully delivers the response to the result router, but the session actor can't correlate it to a client without the registration.

[AI-assisted debugging and comment]
